parser_library/src/lsp/lsp_context.cpp

```diff
@@ -51,470 +52,206 @@ const std::unordered_map<context::section_kind, document_symbol_kind> document_s
-    long long& limit,
-    document_symbol_cache& cache) const
-{
-    std::unordered_map<std::string_view, utils::resource::resource_location> name_to_location;
-    for (const auto& [def, info] : m_macros)
-        name_to_location.insert_or_assign(def->id.to_string_view(), info->definition_location.resource_loc);
-    for (const auto& [def, info] : m_hlasm_ctx->copy_members())
-        name_to_location.insert_or_assign(info->name.to_string_view(), info->definition_location.resource_loc);
+    result.erase(std::unique(result.begin(), result.end(), same_symbol), result.end());
 
-    document_symbol_opencode_ord_symbol(result, limit);
-    document_symbol_opencode_var_seq_symbol_aux(result, name_to_location, limit, cache);
+    static constexpr const auto is_title = [](const auto& s) { return s.kind == TITLE; };
 
-    for (const auto& sym : m_opencode->variable_definitions)
+    for (auto it = result.rbegin(); it != result.rend();)
     {
-        if (limit <= 0)
+        const auto title = std::find_if(it, result.rend(), is_title);
+
+        if (title == result.rend())
             break;
-        if (!belongs_to_copyfile(document_loc, sym.def_position, sym.name))
-        {
-            result.emplace_back(sym.name.to_string(), document_symbol_kind::VAR, range(sym.def_position));
-            --limit;
-        }
-    }
-}
 
-document_symbol_list_s lsp_context::document_symbol(
-    const utils::resource::resource_location& document_loc, long long limit) const
-{
-    document_symbol_list_s result;
-    const auto& file = m_files.find(document_loc);
-    if (file == m_files.end() || limit <= 0)
-        return result;
+        const auto next = std::next(title);
+
+        title->children.insert(
+            title->children.end(), std::make_move_iterator(title.base()), std::make_move_iterator(it.base()));
+        result.erase(title.base(), it.base());
 
-    document_symbol_cache cache;
+        it = next;
+    }
 
-    switch (file->second->type)
+    for (auto it = std::find_if(result.begin(), result.end(), is_title); it != result.end();)
     {
-        case file_type::MACRO:
-            document_symbol_macro(result, document_loc, std::nullopt, limit, cache);
-            break;
+        auto next = std::next(it);
 
-        case file_type::COPY:
-            document_symbol_copy(result, file->second->get_occurrences(), document_loc, std::nullopt, limit);
-            break;
+        if (expand_block(*it))
+        {
+            if (next == result.end())
+            {
+                const auto lines = dl_it->second->data.get_number_of_lines();
+                end_before(it->symbol_range.end, position(lines, 0));
+                end_before(it->symbol_selection_range.end, position(lines, 0));
+            }
+            else
+            {
+                end_before(it->symbol_range.end, next->symbol_range.start);
+                end_before(it->symbol_selection_range.end, next->symbol_range.start);
```

Comment:
Not sure - is this intentional or should it be:

```suggestion
                end_before(it->symbol_selection_range.end, next->symbol_selection_range.start);
```
?